ui/StatusQ/src/writableproxymodel.cpp

```diff
@@ -0,0 +1,743 @@
+        return -1;
+
+    return proxyToSourceRowMapping[row];
+}
+
+int WritableProxyModelPrivate::sourceToProxyRow(int row) const
+{
+    for (int i = 0; i < proxyToSourceRowMapping.size(); ++i) {
+        if(proxyToSourceRowMapping[i] == row)
+            return i;
+    }
+    return -1;
+}
+
+void WritableProxyModelPrivate::createProxyToSourceRowMap()
+{
+    if(!q.sourceModel())
+        return;
+
+    auto sourceModel = q.sourceModel();
+
+    proxyToSourceRowMapping.clear();
+    int sourceIter = 0;
+    for (int i = 0; i < q.rowCount(); ++i) {
+        if(insertedRows.contains(q.index(i, 0)))
+        {
+            proxyToSourceRowMapping.append(-1);
+            continue;
+        }
+        
+        while(removedRows.contains(q.sourceModel()->index(sourceIter, 0)) && sourceIter < sourceModel->rowCount())
+        {
+            ++sourceIter;
+        }
+
+        proxyToSourceRowMapping.append(sourceIter);
+        sourceIter++;
+    }
+}
+
+bool WritableProxyModelPrivate::contains(const QModelIndex& sourceIndex) const {
+    return cache.contains(sourceIndex) || insertedRows.contains(q.mapFromSource(sourceIndex));
+}
+
+void WritableProxyModelPrivate::WritableProxyModelPrivate::clear()
+{
+    cache.clear();
+    removedRows.clear();
+    insertedRows.clear();
+    proxyToSourceRowMapping.clear();
+}
+
+void WritableProxyModelPrivate::clearInvalidatedCache()
+{
+    for(auto iter = removedRows.begin(); iter != removedRows.end();)
+    {
+        if(iter->isValid())
+        {
+            ++iter;
+            continue;
+        }
+
+        iter = removedRows.erase(iter);
+    }
+
+    for(auto iter = cache.begin(); iter != cache.end();)
+    {
+        if(iter.key().isValid())
+        {
+            ++iter;
+            continue;
+        }
+
+        iter = cache.erase(iter);
+    }
+}
+
+void WritableProxyModelPrivate::moveFromCacheToInserted(const QModelIndex& sourceIndex)
+{
+    //User updated this row. Move it in inserted rows. We shouldn't delete it
+    insertedRows.insert(q.mapFromSource(sourceIndex), cache.take(sourceIndex));
+    auto itemData = q.sourceModel()->itemData(sourceIndex);
+    auto proxyIndex = q.mapFromSource(sourceIndex);
+    for(auto it = itemData.begin(); it != itemData.end(); ++it)
+    {
+        if(insertedRows[proxyIndex].contains(it.key()))
+            continue;
+
+        insertedRows[proxyIndex][it.key()] = it.value();
+    }
+}
+
+void WritableProxyModelPrivate::applyRoleNamesFix()
+{
+    if(!q.sourceModel())
+        return;
+    
+    if(q.sourceModel()->rowCount() != 0)
+        return;
+
+    auto connection = std::make_unique<QMetaObject::Connection>();
+    auto connectionPtr = connection.get();
+    
+    auto singleShot = [this, connection = std::move(connection)]() {
+        QObject::disconnect(*connection);
+        q.resetInternalData();
+    };
+    
+    *connectionPtr = q.connect(q.sourceModel(), &QAbstractItemModel::rowsInserted, &q, std::move(singleShot));
```

Comment:
I'm trying really hard to figure out what your aim is here :)

And clang-tidy is warning about a memleak here as well

Reply:
![image](https://github.com/status-im/status-desktop/assets/5377645/e034c728-ed98-407a-96df-00a62605810c)

Reply:
> I'm trying really hard to figure out what your aim is here :)
> 
> And clang-tidy is warning about a memleak here as well

just a single shot connection. I'll have a look

Reply:
When using the ListModel the role names will be added after inserting the first row. In this case we need to reset the internal data. The purpose of this method is to create a single shot connection and reset the internal data on the first insert.

Reply:
The memory leak could happen if the disconnect fails. Not sure yet how to handle this / if we should handle this. I mean we can't just delete a connection because it's managed by Qt.

Reply:
Well, the `disconnect` returns a `bool` so we could handle that

Reply:
That's true, but IDK if there's something we can do since the actual connection is managed by Qt.
The `QMetaObject::Connection` we're keeping here is just a wrapper on top of the real connection object created and managed by Qt. 
The flow should be like this:
1. Create a `unique_ptr<QMetaObject::Connection>` holding the `QMetaObject::Connection` created by QObject::connect
2. Move the unique_ptr in the context of the connection handler.
3. When the connection fires we're disconnecting. When going out of scope, the unique_ptr gets collected and the `QMetaObject::Connection` is released.

Still looking on what I could have missed.

I can change it to something like this. The delete is explicit and hopefully the clang-tidy won't complain:

```
auto connectionPtr = new QMetaObject::Connection()
*connectionPtr = q.connect(q.sourceModel(), &QAbstractItemModel::rowsInserted, &q, [this, connectionPtr](){
      QObject::disconnect(*connection);
       q.resetInternalData();
       delete connectionPtr;
});
```

Reply:
Yeah, let's give it a try

Reply:
Pushed

Reply:
The memory leak warning isn't there anymore. But now it suggests to use a smart ptr instead